File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/controllers/DuplaController.java
package org.example.controllers;

import org.example.models.Dupla;
import org.example.repositories.dupla.DuplaExtension;
import org.example.repositories.dupla.DuplaRepository;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class DuplaController implements DuplaExtension {
    private final DuplaRepository repo;
    public DuplaController(DuplaRepository repo) {
        this.repo = repo;
    }

    @Override
    public List<Dupla> getAll() {
        return repo.getAll();
    }

    @Override
    public Dupla getById(String s) {
        return repo.getById(s);
    }

    @Override
    public Dupla save(Dupla element, boolean storage) {
        return repo.save(element, storage);
    }

    @Override
    public void saveAll(List<Dupla> elements, boolean storage) {
        var paco = elements.stream().filter(Objects::isNull).count();
        repo.saveAll(elements, storage);
    }

    @Override
    public Dupla deleteById(String s) {
        return repo.deleteById(s);
    }

    @Override
    public Dupla update(Dupla element) {
        return repo.update(element);
    }

    @Override
    public Dupla updateById(String s, Dupla element) {
        return repo.updateById(s, element);
    }

    @Override
    public Map<LocalDate, Map<String, Dupla>> maxTemPorLugar() {
        return repo.maxTemPorLugar();
    }

    @Override
    public Map<LocalDate, Map<String, Dupla>> minTemPorLugar() {
        return repo.minTemPorLugar();
    }

    @Override
    public Map<String, Dupla> maxTemPorProvincia() {
        return repo.maxTemPorProvincia();
    }

    @Override
    public Map<String, Dupla> minTemPorProvincia() {
        return repo.minTemPorProvincia();
    }

    @Override
    public Map<String, Double> mediaPorProvincia() {
        return repo.mediaPorProvincia();
    }

    @Override
    public Map<LocalDate, Map<String, Double>> mediaPrecipitacionPorDiaYProvincia() {
        return repo.mediaPrecipitacionPorDiaYProvincia();
    }

    @Override
    public Double mediaTemMadrid() {
        return repo.mediaTemMadrid();
    }

    @Override
    public Double mediaTemMax() {
        return repo.mediaTemMax();
    }

    @Override
    public Double mediaTemMin() {
        return repo.mediaTemMin();
    }

    @Override
    public Map<LocalDate, List<String>> maxTemAntes() {
        return repo.maxTemAntes();
    }

    @Override
    public Map<LocalDate, List<String>> minTemDespues() {
        return repo.minTemDespues();
    }

    @Override
    public Dupla maxTemMadrid(LocalDate day) {
        return repo.maxTemMadrid(day);
    }

    @Override
    public Dupla minTemMadrid(LocalDate day) {
        return repo.minTemMadrid(day);
    }

    @Override
    public Double mediaPrecipitacionMadrid(LocalDate day) {
        return repo.mediaPrecipitacionMadrid(day);
    }
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/controllers/DuplaController.kt
package controllers

import com.github.michaelbull.result.*
import errors.DuplaError
import models.Dupla
import mu.KotlinLogging
import repositories.dupla.DuplaExtension
import repositories.dupla.DuplaRepository
import validators.validate
import java.time.LocalDate

private val logger = KotlinLogging.logger {  }

class DuplaController(
    private val repo: DuplaRepository
) {
    fun maxTemPorLugar(): Map<LocalDate, Map<String, Double>> {
        logger.debug { "DuplaController ->\tmaxTemPorLugar" }
        return repo.maxTemPorLugar()
    }

    fun minTemPorLugar(): Map<LocalDate, Map<String, Double>> {
        logger.debug { "DuplaController ->\tminTemBroupPorLugar" }
        return repo.minTemPorLugar()
    }

    fun maxTemPorProvincia(): Map<String, Double> {
        logger.debug { "DuplaController ->\tmaxTemPorProvincia" }
        return repo.maxTemPorProvincia()
    }

    fun minTemPorProvincia(): Map<String, Double> {
        logger.debug { "DuplaController ->\tminTemPorProvincia" }
        return repo.minTemPorProvincia()
    }

    fun mediaPorProvincia(): Map<String, Double> {
        logger.debug { "DuplaController ->\tmediaPorProvincia" }
        return repo.mediaPorProvincia()
    }

    fun mediaPrecipitacionPorDiaYProvincia(): Map<LocalDate, Map<String, Double>> {
        logger.debug { "DuplaController ->\tmediaPrecipitacionPorDiaYProvincia" }
        return repo.mediaPrecipitacionPorDiaYProvincia()
    }

    fun mediaTemMadrid(): Double {
        logger.debug { "DuplaController ->\tmediaTemMadrid" }
        return repo.mediaTemMadrid()
    }

    fun mediaTemMax(): Double {
        logger.debug { "DuplaController ->\tmediaTemMax" }
        return repo.mediaTemMax()
    }

    fun mediaTemMin(): Double {
        logger.debug { "DuplaController ->\tmediaTemMin" }
        return repo.mediaTemMin()
    }

    fun maxTemAntes(): Map<LocalDate, List<String>> {
        logger.debug { "DuplaController ->\tmaxTemAntes" }
        return repo.maxTemAntes()
    }

    fun minTemDespues(): Map<LocalDate, List<String>> {
        logger.debug { "DuplaController ->\tminTemDespues" }
        return repo.minTemDespues()
    }

    fun maxTemMadrid(day: LocalDate): Dupla {
        logger.debug { "DuplaController ->\tmaxTemMadrid" }
        return repo.maxTemMadrid(day)
    }

    fun minTemMadrid(day: LocalDate): Dupla {
        logger.debug { "DuplaController ->\tminTemMadrid" }
        return repo.minTemMadrid(day)
    }

    fun mediaPrecipitacionMadrid(day: LocalDate): Double {
        logger.debug { "DuplaController ->\tmediaPrecipitacionMadrid" }
        return repo.mediaPrecipitacionMadrid(day)
    }

    fun getAll(): Iterable<Dupla> {
        logger.debug { "DuplaController ->\tgetAll" }
        return repo.getAll()
    }

    fun getById(id: String): Result<Dupla, DuplaError> {
        logger.debug { "DuplaController ->\tgetById" }
        return repo.getById(id)?.let { Ok(it) } ?: Err(DuplaError.DuplaNoEncontradaError())
    }

    fun save(element: Dupla, storage: Boolean = true): Result<Dupla, DuplaError> {
        logger.debug { "DuplaController ->\tsave" }
        return element.validate().onSuccess { repo.save(element, storage) }
    }

    fun saveAll(elements: Iterable<Dupla>, storage: Boolean = true) {
        logger.debug { "DuplaController ->\tsaveAll" }
        repo.saveAll(elements, storage)
    }

    fun deleteById(id: String): Result<Dupla, DuplaError> {
        logger.debug { "DuplaController ->\tdeleteById" }
        return repo.deleteById(id)?.let { Ok(it) } ?: Err(DuplaError.DuplaNoEncontradaError())
    }

    fun update(element: Dupla): Result<Dupla, DuplaError> {
        logger.debug { "DuplaController ->\tupdate" }
        return element.validate().andThen { repo.update(element)?.let { Ok(it) } ?: Err(DuplaError.DuplaNoEncontradaError()) }
    }

    fun updateById(id: String, element: Dupla): Result<Dupla, DuplaError> {
        logger.debug { "DuplaController ->\tupdateById" }
        return element.validate().andThen { repo.updateById(id, element)?.let { Ok(it) } ?: Err(DuplaError.DuplaNoEncontradaError()) }
    }
}